mautic/mautic
# Use message queue protocol to handle email tracking.

Please answer the following questions. 

| Q | A |
| --- | --- |
| Bug fix? | X |
| New feature? | Yes |
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/122 |
| Related developer documentation PR URL | X |
| Issues addressed (#s or URLs) | X |
| BC breaks? | X |
| Deprecations? | X |
### Required
#### Description:

We are having problems with big emails campaigns because it caused a lock in our database when too many emails are opened simultaneously.

We add the ability to use a message queue protocol - now only implemented is [rabbitmq](https://www.rabbitmq.com/).

We can now postpone tracking emails orders and only consume 50 per minute for example.

You can configure emails in the email section of the administration panel.

<img width="854" alt="admin" src="https://cloud.githubusercontent.com/assets/2357779/17698946/42abf8be-63be-11e6-8f83-20502cfa5b6b.png">

We are using the [RabbitMq Bundle](https://github.com/php-amqplib/RabbitMqBundle) which operates nicely with symfony and offer a wide range of messaging patterns. 

If you want to use it in a production or staging environment i recommand to use [cloudamqp](https://www.cloudamqp.com/) which offers rabbit mq as a service ( the free plan should be sufficient at first ).
#### Call to discussions

One of the goals of this PR is to know if you are facing the same problems and discuss on how to handle mautic scalability.

Having more than one thousand of segments, shootings emails happens from time to time and you have be sure that your infra can handle when those emails are to be opened.

Either you have a suboptimize infra or an autoscaling one but it costs a lot.

It is the same problem for tracking pixel, but the code is more intricate and we were not able to make it work.

So do you think that message queue can help tackle Mautic's scalability ?
#### Steps to test this PR:
1. You should first install a rabbitmq client, go to the [download page](https://www.rabbitmq.com/download.html)
2. Configure everything related in `rabbitmq.yml` ( there is the default config for testing )
3. Launched the rabbitmq server : `rabbitmq-server [-detached]`
4. Go to the admin page and change the option to true ( default is false )
5. Init the system by calling : `./app/console rabbitmq:setup-fabric`, this will create the exchange and the queue. ( this step is required, but you can do it manually )
6. Call the url that simulated an open email : `mautic_url/email/{id}.gif`
7. Check that a message has been pushed to the queue ( should be `task_email` by default )

Later on you can consume the message with this commande : `./app/console rabbitmq:consumer task_email`, we can add many options like `-m 5` to consume only 5 messages.
